Get started docs (#15098)
huggingface/transformers: docs/source/quicktour.mdx

not, the code is expected to work for both backends without any change.
</Tip>

## Pipeline

[`pipeline`] is the easiest way to use a pretrained model for a given task.

<Youtube id="tiZFewofSLM"/>

The [`pipeline`] supports many common tasks out-of-the-box:

**Text**:
* Sentiment analysis: classify the polarity of a given text.
* Text generation (in English): generate text from a given input.
* Name entity recognition (NER): label each word with the entity it represents (person, date, location, etc.).
* Question answering: extract the answer from the context, given some context and a question.
* Fill-mask: fill in the blank given a text with masked words.
* Summarization: generate a summary of a long sequence of text or document.
* Translation: translate text into another language.
* Feature extraction: create a tensor representation of the text.

**Image**:
* Image classification: classify an image.
* Image segmentation: classify every pixel in an image.
* Object detection: detect objects within an image.

**Audio**:
* Audio classification: assign a label to a given segment of audio.

> * Automatic speech recognition (ASR): transcribe audio data into text.

Reviewer: Great to have all the modalities presented!

<Tip>
For more details about the [`pipeline`] and associated tasks, refer to the documentation [here](./main_classes/pipelines)
</Tip>

### Using the pipeline

In the following example, we will look at how to use the [`pipeline`] for sentiment analysis.

Install the following dependencies if you haven't already:

```bash
pip install torch
===PT-TF-SPLIT===
pip install tensorflow
```

Import [`pipeline`] and specify the task you want to complete:

```py
>>> from transformers import pipeline

>>> classifier = pipeline("sentiment-analysis")
```

The pipeline downloads and caches a default [pretrained model](https://huggingface.co/distilbert-base-uncased-finetuned-sst-2-english) and tokenizer for sentiment analysis. Now you can use the `classifier` on your target text:

```py
>>> classifier("We are very happy to show you the 🤗 Transformers library."")
[{"label": "POSITIVE", "score": 0.9998}]
```

For more than one sentence, pass a list of sentences to the [`pipeline`] which returns a list of dictionaries:

```py
>>> results = classifier(["We are very happy to show you the 🤗 Transformers library.", "We hope you don't hate it."])
>>> for result in results:

> ...     print(f"label: {result["label"]}, with score: {round(result["score"], 4)}")

Reviewer 2: I think this change is breaking the code quality script, could you revert it? You shouldn't use `"` inside an f-string which uses `"` to define the string.

label: POSITIVE, with score: 0.9998
label: NEGATIVE, with score: 0.5309
```

The [`pipeline`] can also iterate over an entire dataset. Start by installing the 🤗 [Datasets](https://huggingface.co/docs/datasets/) library:

```bash
pip install datasets 
```

Create a [`pipeline`] with the task you want to solve for and the model you want to use. Set the `device` parameter to `0` to place the tensors on a CUDA device:

```py
>>> from transformers import pipeline

>>> speech_recognizer = pipeline("automatic-speech-recognition", model="facebook/wav2vec2-base-960h", device=0)
```

> Next, load a dataset (see the 🤗 Datasets [Quick Start](https://huggingface.co/docs/datasets/quickstart.html#) for more details) you'd like to iterate over. For example, let's load the [SUPERB](https://huggingface.co/datasets/superb) dataset:

Reviewer: ```suggestion
Next, load a dataset (see the 🤗 Datasets [Quick Start](https://huggingface.co/docs/datasets/quickstart.html) for more details) you'd like to iterate over. For example, let's load the [SUPERB](https://huggingface.co/datasets/superb) dataset:
```

```py
>>> import datasets

>>> dataset = datasets.load_dataset("superb", name="asr", split="test")
```

Now you can iterate over the dataset with the pipeline. `KeyDataset` retrieves the item in the dictionary returned by the dataset:

```py
>>> for out in tqdm.tqdm(speech_recognizer(KeyDataset(dataset, "file"))):
        print(out)
        # {"text": "HE HOPED THERE WOULD BE STEW FOR DINNER TURNIPS AND CARROTS AND BRUISED POTATOES AND FAT MUTTON PIECES TO BE LADLED OUT IN THICK PEPPERED FLOWER FAT AND SAUCE"}
        # {"text": ....}

>         # ....

Reviewer: This shouldn't be set as comments here, we can use another code block to show the first output.

```

### Use another model and tokenizer in the pipeline

The [`pipeline`] can accommodate any model from the Model Hub, making it easy to adapt the [`pipeline`] for other use-cases. For example, if you'd like a model capable of handling French text, use the tags on the [Model Hub](https://huggingface.co/models) to filter for an appropriate model. The top filtered result returns a multilingual [BERT model](https://huggingface.co/nlptown/bert-base-multilingual-uncased-sentiment) fine-tuned for sentiment analysis. Great, let's use this model!

```py
>>> model_name = "nlptown/bert-base-multilingual-uncased-sentiment"
```

> Use the [`AutoModel`] and ['AutoTokenizer'] to load the pretrained model and it's associated tokenizer (more on the `AutoClasses` below):

Reviewer: ```suggestion
Use the [`AutoModelForSequenceClassification`] and [`AutoTokenizer`] to load the pretrained model and it's associated tokenizer (more on the `AutoClasses` below):
```

>>> model = AutoModelForSequenceClassification.from_pretrained(model_name)
>>> tokenizer = AutoTokenizer.from_pretrained(model_name)
===PT-TF-SPLIT===
>>> from transformers import AutoTokenizer, TFAutoModelForSequenceClassification

>>> model = TFAutoModelForSequenceClassification.from_pretrained(model_name, from_pt=True)
>>> tokenizer = AutoTokenizer.from_pretrained(model_name)
```

Then you can specify the model and tokenizer in the [`pipeline`], and apply the `classifier` on your target text:

```py
>>> classifier = pipeline('sentiment-analysis', model=model, tokenizer=tokenizer)
>>> classifier("Nous sommes très heureux de vous présenter la bibliothèque 🤗 Transformers.")
[{"label": "5 stars", "score": 0.7272651791572571}]
```

If you can't find a model for your use-case, you will need to fine-tune a pretrained model on your data. Take a look at our [fine-tuning tutorial](./training) to learn how. Finally, after you've fine-tuned your pretrained model, please consider sharing it (see tutorial [here](./model_sharing)) with the community on the Model Hub to democratize NLP for everyone! 🤗

## AutoClasses

<Youtube id="AhChOFRegn4"/>

> Under the hood, the [`AutoModel`] and [`AutoTokenizer`] classes work together to power the [`pipeline`]. The [`AutoClasses`](./model_doc/auto) are shortcuts that automatically retrieves the model architecture of a pretrained model from it's name or path. You only need to select the appropriate [`AutoModel`] for your task and it's associated tokenizer with [`AutoTokenizer`]. 

Reviewer: ```suggestion
Under the hood, the [`AutoModelForSequenceClassification`] and [`AutoTokenizer`] classes work together to power the [`pipeline`]. The [`AutoClasses`](./model_doc/auto) are shortcuts that automatically retrieves the model architecture of a pretrained model from it's name or path. You only need to select the appropriate auto class for your task and it's associated tokenizer with [`AutoTokenizer`]. 
```

Let's return to our example and see how we can use the `AutoClasses` to replicate the results of the [`pipeline`].

### Using the AutoTokenizer

A tokenizer is responsible for preprocessing text in a way that is understandable by the model. First, the tokenizer will split a given text into words called *tokens*. There are multiple rules that govern the tokenization process, including how to split a word and at what level (learn more [here](./tokenizer_summary)). The most important thing to remember though is you need to instantiate the tokenizer with the name of the model to ensure you're using the same tokenization rules a model was pretrained with.

Load a tokenizer with [`AutoTokenizer`]:

```py
>>> from transformers import AutoTokenizer, AutoModelForSequenceClassification

>>> model_name = "distilbert-base-uncased-finetuned-sst-2-english"

> >>> pt_model = AutoModelForSequenceClassification.from_pretrained(model_name)

Reviewer: If we don't load the model in this code sample, let's remove the framework switch and the imports.

> * [atttention_mask](.glossary#attention-mask): indicates which tokens should be attended to.

Reviewer: ```suggestion
* [atttention_mask](glossary#attention-mask): indicates which tokens should be attended to.
```
(that or ./ but not just .)

... )
```

Read the [preprocessing](./preprocessing) tutorial for more details about tokenization.

### Using the AutoModel

> 🤗 Transformers provides a simple and unified way to load pretrained instances, which means you can load an [`AutoModel`] like you would load an [`AutoTokenizer`]. The only difference is selecting the correct [`AutoModel`] for the task. Since we are doing text - or sequence - classification, load [`AutoModelForSequenceClassification`]. The TensorFlow equivalent is simply [`TFAutoModelForSequenceClassification`]:

Reviewer: I would use `AutoClass` instead of `AutoModel`, as `AutoModel` exists.

```py

> >>> model_name = "nlptown/bert-base-multilingual-uncased-sentiment"

Reviewer: The imports removed above would go here.

>>> pt_model = AutoModelForSequenceClassification.from_pretrained(model_name)
===PT-TF-SPLIT===
>>> model_name = "nlptown/bert-base-multilingual-uncased-sentiment"

> >>> # This model only exists in PyTorch, so we use the _from_pt_ flag to import that model in TensorFlow.

Reviewer: @LysandreJik Can we upload the TF weights of this model on the repo? It would be great to have the doc example work out of the box.

Reviewer 2: Yes, but this will take a bit of time while the authors respond. 

I think it's fine to not slow the PR merge because of this, I'm contacting the authors straight away. (From my side: feel free to merge the PR like this, but I'll let you decide @sgugger)

Reviewer 2: Authors contacted, checkpoint uploaded, feel free to merge

>>> from torch import nn

>>> pt_predictions = nn.functional.softmax(pt_outputs.logits, dim=-1)
>>> print(pt_predictions)
tensor([[2.2043e-04, 9.9978e-01],
        [5.3086e-01, 4.6914e-01]], grad_fn=<SoftmaxBackward>)
===PT-TF-SPLIT===
>>> import tensorflow as tf

>>> tf_predictions = tf.nn.softmax(tf_outputs.logits, axis=-1)
>>> print(tf_predictios)
tf.Tensor(
[[2.2043e-04 9.9978e-01]
 [5.3086e-01 4.6914e-01]], shape=(2, 2), dtype=float32)
```

<Tip>
All 🤗 Transformers models (PyTorch or TensorFlow) outputs the tensors *before* the final activation
function (like softmax) because the final activation function is often fused with the loss.
</Tip>

Models are standard [`torch.nn.Module`](https://pytorch.org/docs/stable/nn.html#torch.nn.Module) or [tf.keras.Model](https://www.tensorflow.org/api_docs/python/tf/keras/Model) so you can use them in your usual training loop. However, to make things easier, 🤗 Transformers provides a [`Trainer`] class for PyTorch that adds functionality for distributed training, mixed precision, and more. For TensorFlow, you can use the `fit()` method from [Keras](https://keras.io/). Refer to the [training tutorial](./training) for more details.

<Tip>

> PyTorch model outputs are special dataclasses so their attributes are autocompleted in an IDE.

Reviewer: ```suggestion
🤗 Transformers model outputs are special dataclasses so their attributes are autocompleted in an IDE.
```
It's not a PyTorch (or TF) thing.
